File: server/review_building.py
# server/review_building.py
from flask import Blueprint, jsonify, request

import server.queries.review_building as queries
import server.services.review_building as services

review_building_bp = Blueprint(
    "review_building", __name__, url_prefix="/api/review-building"
)


@review_building_bp.get("/next")
def next_building():
    return jsonify({"row_id": queries.get_first_pending_building_id()})


@review_building_bp.get("/<int:post_id>")
def get_building(post_id):
    result = services.get_review_building_with_nav(post_id)
    if result is None:
        return jsonify({"error": "not found"}), 404
    return jsonify(result)


@review_building_bp.post("/<int:post_id>/enter")
def enter(post_id):
    try:
        result = services.enter(post_id)
        return jsonify({"ok": True, **result})
    except LookupError as e:
        return jsonify({"error": str(e)}), 404


@review_building_bp.post("/<int:post_id>/assign")
def assign(post_id):
    building_id = request.json.get("building_id")
    if building_id is None:
        return jsonify({"error": "Missing building_id"}), 400
    try:
        result = services.assign(post_id, int(building_id))
        return jsonify({"ok": True, **result})
    except LookupError as e:
        return jsonify({"error": str(e)}), 404


@review_building_bp.post("/<int:post_id>/dismiss")
def dismiss(post_id):
    try:
        services.dismiss(post_id)
        return jsonify({"ok": True})
    except LookupError as e:
        return jsonify({"error": str(e)}), 404

File: frontend/src/api/reviewBuilding.ts
// frontend/src/api/reviewBuilding.ts
import type { Post } from "./posts";
import { createRequest } from "./http";

const request = createRequest("/api/review-building");

export interface BuildingOption {
  id: number;
  name: string;
}

export interface NavContext {
  current: number;
  total: number;
  prev_id: number | null;
  next_id: number | null;
}

export interface BuildingReviewWithNav {
  post: Post;
  nav: NavContext;
  buildings: BuildingOption[];
}

export function getNextBuildingReview(): Promise<{ row_id: number | null }> {
  return request("/next");
}

export function getBuildingReview(
  postId: number,
): Promise<BuildingReviewWithNav> {
  return request(`/${postId}`);
}

export type EnterBuildingResult =
  | { ok: true; property_id: number | null }
  | { ok: true; sent_to_review: true; candidate_property_id: number | null }
  | { ok: true; discarded: true };

export function enterBuilding(postId: number): Promise<EnterBuildingResult> {
  return request(`/${postId}/enter`, { method: "POST" });
}

export function assignBuilding(
  postId: number,
  buildingId: number,
): Promise<{ ok: boolean; property_id: number | null }> {
  return request(`/${postId}/assign`, {
    method: "POST",
    body: JSON.stringify({ building_id: buildingId }),
  });
}

export function dismissBuildingReview(
  postId: number,
): Promise<{ ok: boolean }> {
  return request(`/${postId}/dismiss`, { method: "POST" });
}
